File: src-tauri/src/simulations/shared/camera.rs
use super::coordinates::{CoordinateTransform, NdcCoords, ScreenCoords, WorldCoords};
use crate::error::SimulationResult;
use bytemuck::{Pod, Zeroable};
use serde_json;
use std::sync::Arc;
use wgpu::{Device, Queue};

/// GPU-compatible camera uniform data
#[repr(C)]
#[derive(Copy, Clone, Pod, Zeroable, Debug)]
pub struct CameraUniform {
    /// Simple 2D transformation matrix (4x4 matrix stored as 16 floats)
    pub transform_matrix: [f32; 16],
    /// Camera position in world space
    pub position: [f32; 2],
    /// Zoom level (scale factor)
    pub zoom: f32,
    /// Aspect ratio (width/height)
    pub aspect_ratio: f32,
}

impl CoordinateTransform for Camera {
    fn screen_to_world(&self, screen: ScreenCoords) -> WorldCoords {
        self.screen_to_world(screen)
    }

    fn world_to_screen(&self, world: WorldCoords) -> ScreenCoords {
        self.world_to_screen(world)
    }

    fn screen_to_ndc(&self, screen: ScreenCoords) -> NdcCoords {
        self.screen_to_ndc(screen)
    }

    fn ndc_to_world(&self, ndc: NdcCoords) -> WorldCoords {
        self.ndc_to_world(ndc)
    }

    fn world_to_ndc(&self, world: WorldCoords) -> NdcCoords {
        self.world_to_ndc(world)
    }
}

/// Simple 2D camera system for GPU rendering
#[derive(Debug)]
pub struct Camera {
    /// Camera position in world space
    pub position: [f32; 2],
    /// Target camera position for smooth movement
    target_position: [f32; 2],
    /// Zoom level (scale factor)
    pub zoom: f32,
    /// Target zoom level for smooth zooming
    target_zoom: f32,
    /// Viewport dimensions
    pub viewport_width: f32,
    pub viewport_height: f32,
    /// GPU buffer for camera uniform data
    buffer: wgpu::Buffer,
    /// Cached uniform data
    uniform_data: CameraUniform,
    /// Smoothing factor for camera movement (0.0 = no smoothing, 1.0 = instant)
    smoothing_factor: f32,
    /// Camera sensitivity multiplier for pan and zoom operations
    sensitivity: f32,
    /// Optional position clamp; when None, panning is unbounded
    position_clamp: Option<(f32, f32)>,
}

impl Camera {
    /// Create a new camera with default settings
    pub fn new(
        device: &Arc<Device>,
        viewport_width: f32,
        viewport_height: f32,
    ) -> SimulationResult<Self> {
        let position = [0.0, 0.0]; // Center of [-1,1] space
        let zoom = 1.0; // No zoom
        let aspect_ratio = viewport_width / viewport_height;

        let uniform_data = CameraUniform {
            transform_matrix: Self::create_simple_transform_matrix(position, zoom),
            position,
            zoom,
            aspect_ratio,
        };

        let buffer = device.create_buffer(&wgpu::BufferDescriptor {
            label: Some("Camera Uniform Buffer"),
            size: std::mem::size_of::<CameraUniform>() as u64,
            usage: wgpu::BufferUsages::UNIFORM | wgpu::BufferUsages::COPY_DST,
            mapped_at_creation: false,
        });

        Ok(Self {
            position,
            target_position: position,
            zoom,
            target_zoom: zoom,
            viewport_width,
            viewport_height,
            buffer,
            uniform_data,
            smoothing_factor: 0.15, // Smooth camera movement
            sensitivity: 1.0,       // Default sensitivity
            position_clamp: Some((-2.0, 2.0)),
        })
    }

    /// Create a simple 2D transformation matrix
    fn create_simple_transform_matrix(position: [f32; 2], zoom: f32) -> [f32; 16] {
        // Create a simple orthographic projection matrix
        // This maps [-1,1] x [-1,1] world space to [-1,1] x [-1,1] clip space
        let scale_x = zoom;
        let scale_y = zoom;

        // For proper center zooming, we want to:
        // 1. Scale around the origin (0,0)
        // 2. Then translate to account for camera position
        // The translation should move the camera center to NDC origin (0,0)
        let translate_x = -position[0] * zoom;
        let translate_y = -position[1] * zoom;

        [
            scale_x,
            0.0,
            0.0,
            0.0,
            0.0,
            scale_y,
            0.0,
            0.0,
            0.0,
            0.0,
            1.0,
            0.0,
            translate_x,
            translate_y,
            0.0,
            1.0,
        ]
    }

    /// Update camera state (call this every frame for smooth movement)
    pub fn update(&mut self, delta_time: f32) -> bool {
        // Apply smoothing to position
        let smoothing = self.smoothing_factor * delta_time * 60.0; // Adjust for frame rate
        let smoothing = smoothing.min(1.0); // Clamp to prevent overshooting

        self.position[0] += (self.target_position[0] - self.position[0]) * smoothing;
        self.position[1] += (self.target_position[1] - self.position[1]) * smoothing;

        // Apply smoothing to zoom
        self.zoom += (self.target_zoom - self.zoom) * smoothing;

        // Update uniform data after smoothing
        self.update_uniform();
        true
    }

    /// Update camera position (panning)
    pub fn pan(&mut self, delta_x: f32, delta_y: f32) {
        let pan_speed = 0.1 / self.zoom; // Pan speed depends on zoom level

        // Apply sensitivity to the pan movement
        let adjusted_delta_x = delta_x * self.sensitivity;
        let adjusted_delta_y = delta_y * self.sensitivity;

        // Update target position instead of current position
        self.target_position[0] += adjusted_delta_x * pan_speed;
        self.target_position[1] += adjusted_delta_y * pan_speed;

        if let Some((min, max)) = self.position_clamp {
            self.target_position[0] = self.target_position[0].clamp(min, max);
            self.target_position[1] = self.target_position[1].clamp(min, max);
        }
    }

    /// Update zoom level (zooms to center of viewport)
    pub fn zoom(&mut self, delta: f32) {
        // Apply sensitivity to the zoom movement
        let adjusted_delta = delta * self.sensitivity;
        let zoom_factor = 1.0 + adjusted_delta * 0.3;
        let new_zoom = self.target_zoom * zoom_factor;
        let clamped_zoom = new_zoom.clamp(0.005, 50.0); // Set reasonable minimum zoom level

        // Use relative threshold to handle extreme zoom levels properly
        let relative_threshold = self.target_zoom * 0.001; // 0.1% change threshold
        let absolute_threshold = 0.000001; // Much smaller minimum absolute change for extreme zoom levels
        let threshold = relative_threshold.max(absolute_threshold);

        if (clamped_zoom - self.target_zoom).abs() > threshold {
            // Store the center point in world coordinates before zoom
            let center_world_x = self.target_position[0];
            let center_world_y = self.target_position[1];

            self.target_zoom = clamped_zoom;

            // Keep the same center point after zoom
            self.target_position[0] = center_world_x;
            self.target_position[1] = center_world_y;
        }
    }

    /// Zoom towards a specific screen position (for mouse wheel)
    pub fn zoom_to_cursor(&mut self, delta: f32, cursor_x: f32, cursor_y: f32) {
        let old_zoom = self.target_zoom;
        self.zoom(delta);

        // Convert cursor position to NDC coordinates
        let mouse_x_norm = (cursor_x / self.viewport_width) * 2.0 - 1.0;
        let mouse_y_norm = -((cursor_y / self.viewport_height) * 2.0 - 1.0);

        // Calculate the world point under the cursor before zoom (use target position for consistency)
        let world_x = mouse_x_norm / old_zoom + self.target_position[0];
        let world_y = mouse_y_norm / old_zoom + self.target_position[1];

        // Calculate where this world point should be in NDC after zoom
        let new_ndc_x = (world_x - self.target_position[0]) * self.target_zoom;
        let new_ndc_y = (world_y - self.target_position[1]) * self.target_zoom;

        // Calculate the offset needed to keep the cursor point stationary
        let offset_x = (mouse_x_norm - new_ndc_x) / self.target_zoom;
        let offset_y = (mouse_y_norm - new_ndc_y) / self.target_zoom;

        // Apply the offset to keep the cursor point stationary
        self.target_position[0] += offset_x;
        self.target_position[1] += offset_y;

        if let Some((min, max)) = self.position_clamp {
            self.target_position[0] = self.target_position[0].clamp(min, max);
            self.target_position[1] = self.target_position[1].clamp(min, max);
        }
    }

    /// Reset camera to default position and zoom
    pub fn reset(&mut self) {
        self.position = [0.0, 0.0];
        self.target_position = [0.0, 0.0];
        self.zoom = 1.0;
        self.target_zoom = 1.0;
        self.update_uniform();
    }

    /// Update viewport dimensions (call when window is resized)
    pub fn resize(&mut self, width: f32, height: f32) {
        self.viewport_width = width;
        self.viewport_height = height;
        self.update_uniform();
    }

    /// Update the uniform data after camera changes
    fn update_uniform(&mut self) {
        let aspect_ratio = self.viewport_width / self.viewport_height;
        self.uniform_data = CameraUniform {
            transform_matrix: Self::create_simple_transform_matrix(self.position, self.zoom),
            position: self.position,
            zoom: self.zoom,
            aspect_ratio,
        };
    }

    /// Upload camera data to GPU buffer
    pub fn upload_to_gpu(&self, queue: &Queue) {
        queue.write_buffer(&self.buffer, 0, bytemuck::cast_slice(&[self.uniform_data]));
    }

    /// Get the GPU buffer for binding to render passes
    pub fn buffer(&self) -> &wgpu::Buffer {
        &self.buffer
    }

    /// Convert screen coordinates to world coordinates
    pub fn screen_to_world(&self, screen: ScreenCoords) -> WorldCoords {
        let ndc_x = (screen.x / self.viewport_width) * 2.0 - 1.0;
        let ndc_y = -((screen.y / self.viewport_height) * 2.0 - 1.0); // Flip Y axis correctly

        // Use target position and zoom for immediate response (no smoothing lag)
        let world_x = (ndc_x / self.target_zoom) + self.target_position[0];
        let world_y = (ndc_y / self.target_zoom) + self.target_position[1];

        WorldCoords::new(world_x, world_y)
    }

    /// Convert world coordinates to screen coordinates
    pub fn world_to_screen(&self, world: WorldCoords) -> ScreenCoords {
        let ndc_x = (world.x - self.target_position[0]) * self.target_zoom;
        let ndc_y = (world.y - self.target_position[1]) * self.target_zoom;

        let screen_x = (ndc_x + 1.0) * self.viewport_width * 0.5;
        let screen_y = (-ndc_y + 1.0) * self.viewport_height * 0.5; // Flip Y axis correctly

        ScreenCoords::new(screen_x, screen_y)
    }

    /// Convert screen coordinates to NDC
    pub fn screen_to_ndc(&self, screen: ScreenCoords) -> NdcCoords {
        let ndc_x = (screen.x / self.viewport_width) * 2.0 - 1.0;
        let ndc_y = -((screen.y / self.viewport_height) * 2.0 - 1.0); // Flip Y axis correctly
        NdcCoords::new(ndc_x, ndc_y)
    }

    /// Convert NDC to screen coordinates
    pub fn ndc_to_screen(&self, ndc: NdcCoords) -> ScreenCoords {
        let screen_x = (ndc.x + 1.0) * self.viewport_width * 0.5;
        let screen_y = (-ndc.y + 1.0) * self.viewport_height * 0.5; // Flip Y axis correctly
        ScreenCoords::new(screen_x, screen_y)
    }

    /// Convert NDC to world coordinates
    pub fn ndc_to_world(&self, ndc: NdcCoords) -> WorldCoords {
        let world_x = (ndc.x / self.target_zoom) + self.target_position[0];
        let world_y = (ndc.y / self.target_zoom) + self.target_position[1];
        WorldCoords::new(world_x, world_y)
    }

    /// Convert world coordinates to NDC
    pub fn world_to_ndc(&self, world: WorldCoords) -> NdcCoords {
        let ndc_x = (world.x - self.target_position[0]) * self.target_zoom;
        let ndc_y = (world.y - self.target_position[1]) * self.target_zoom;
        NdcCoords::new(ndc_x, ndc_y)
    }

    /// Get camera state for frontend serialization
    pub fn get_state(&self) -> serde_json::Value {
        serde_json::json!({
            "position": [self.position[0], self.position[1], 0.0],
            "zoom": self.zoom,
            "viewport_width": self.viewport_width,
            "viewport_height": self.viewport_height,
            "aspect_ratio": self.viewport_width / self.viewport_height
        })
    }

    /// Set the camera smoothing factor
    pub fn set_smoothing_factor(&mut self, factor: f32) {
        self.smoothing_factor = factor.clamp(0.0, 1.0);
    }

    /// Get the current camera smoothing factor
    pub fn get_smoothing_factor(&self) -> f32 {
        self.smoothing_factor
    }

    /// Get the target camera position
    pub fn get_target_position(&self) -> [f32; 2] {
        self.target_position
    }

    /// Get the target zoom level
    pub fn get_target_zoom(&self) -> f32 {
        self.target_zoom
    }

    /// Set the camera sensitivity
    pub fn set_sensitivity(&mut self, sensitivity: f32) {
        self.sensitivity = sensitivity.clamp(0.1, 5.0);
    }

    /// Get the current camera sensitivity
    pub fn get_sensitivity(&self) -> f32 {
        self.sensitivity
    }

    /// Set position clamp for panning. Pass None for unbounded panning.
    pub fn set_position_clamp(&mut self, clamp: Option<(f32, f32)>) {
        self.position_clamp = clamp;
    }
}

File: src/lib/engine/core/Camera.ts
/**
 * 2D camera.
 *
 * Port of src-tauri/src/simulations/shared/camera.rs (369 ln).
 *
 * Two positions and two zooms are tracked: `target*` is what input writes and
 * what every coordinate conversion reads, `current` is the exponentially
 * smoothed value that actually reaches the GPU. Conversions deliberately read
 * the target so a click lands where the user pointed rather than where the
 * smoothing happens to be mid-flight — the Rust makes the same choice and says
 * so at camera.rs:278.
 *
 * The transform is `ndc = (world - position) * zoom`, with no aspect-ratio term.
 * NDC therefore spans [-1,1] on both axes regardless of viewport shape, which
 * means world units are non-square on a non-square viewport. That is what the
 * shaders expect (`CameraUniform.aspect_ratio` is uploaded but the matrix does
 * not use it), so it is preserved rather than "fixed".
 *
 * No GPU object is created at construction, so this module imports and
 * instantiates fine in node with no WebGPU. Call `attachToDevice()` when there
 * is a device.
 */

import type { CameraState } from '../types';
import {
    clamp,
    ndcCoords,
    screenCoords,
    worldCoords,
    type CoordinateTransform,
    type NdcCoords,
    type ScreenCoords,
    type WorldCoords,
} from './coordinates';

/** Every magic number in camera.rs, named. */
export const CAMERA_DEFAULTS = {
    /** camera.rs:104 — fraction of the remaining distance closed per 1/60 s. */
    smoothingFactor: 0.15,
    /** camera.rs:105 */
    sensitivity: 1.0,
    /** camera.rs:106 — panning is bounded to this box in world space. */
    positionClamp: [-2.0, 2.0] as [number, number],
    /** camera.rs:185 */
    minZoom: 0.005,
    /** camera.rs:185 */
    maxZoom: 50.0,
    /** camera.rs:163 — pan step is `panSpeed / zoom`, so panning slows as you zoom in. */
    panSpeed: 0.1,
    /** camera.rs:183 — one wheel notch of delta 1.0 scales zoom by 1.3. */
    zoomStep: 0.3,
    /** camera.rs:147 — smoothing is expressed per 60 Hz frame, hence the ×60. */
    smoothingReferenceHz: 60.0,
    /** camera.rs:188 — a zoom change smaller than 0.1% of current zoom is ignored. */
    zoomRelativeThreshold: 0.001,
    /** camera.rs:189 — floor on the above, so extreme zoom levels still respond. */
    zoomAbsoluteThreshold: 0.000001,
    /** camera.rs:337 */
    smoothingRange: [0.0, 1.0] as [number, number],
    /** camera.rs:357 */
    sensitivityRange: [0.1, 5.0] as [number, number],
} as const;

/** `CameraUniform` is 16 matrix floats + vec2 position + zoom + aspect = 20 floats. */
export const CAMERA_UNIFORM_FLOATS = 20;
export const CAMERA_UNIFORM_BYTES = CAMERA_UNIFORM_FLOATS * 4;

/** The shape `get_camera_state` returned from Rust (camera.rs:325). */
export interface CameraStateJson {
    /** Three elements — the Rust padded a 2D position to 3. */
    position: [number, number, number];
    zoom: number;
    viewport_width: number;
    viewport_height: number;
    aspect_ratio: number;
}

export class Camera implements CoordinateTransform {
    /** Smoothed position — what the GPU sees. */
    position: [number, number] = [0, 0];
    /** Smoothed zoom — what the GPU sees. */
    zoom = 1;

    private targetPosition: [number, number] = [0, 0];
    private targetZoom = 1;

    viewportWidth: number;
    viewportHeight: number;

    private smoothingFactor: number = CAMERA_DEFAULTS.smoothingFactor;
    private sensitivity: number = CAMERA_DEFAULTS.sensitivity;
    private positionClamp: [number, number] | null = [...CAMERA_DEFAULTS.positionClamp];

    private buffer: GPUBuffer | null = null;

    /**
     * @param viewportWidth  canvas **backing-store** width (`canvas.width`)
     * @param viewportHeight canvas **backing-store** height (`canvas.height`)
     */
    constructor(viewportWidth: number, viewportHeight: number) {
        this.viewportWidth = viewportWidth;
        this.viewportHeight = viewportHeight;
    }

    // -----------------------------------------------------------------------
    // Frame update
    // -----------------------------------------------------------------------

    /**
     * Advance the smoothing. Call once per rendered frame.
     *
     * camera.rs:145. The factor is scaled by `dt * 60` so the feel is
     * frame-rate independent, then clamped to 1 so a long frame (a backgrounded
     * tab, a stalled pipeline compile) snaps to the target instead of
     * overshooting past it and oscillating.
     */
    update(deltaTime: number): boolean {
        const smoothing = Math.min(
            this.smoothingFactor * deltaTime * CAMERA_DEFAULTS.smoothingReferenceHz,
            1.0
        );

        this.position[0] += (this.targetPosition[0] - this.position[0]) * smoothing;
        this.position[1] += (this.targetPosition[1] - this.position[1]) * smoothing;
        this.zoom += (this.targetZoom - this.zoom) * smoothing;

        return true;
    }

    /** True while the smoothed values have not yet caught up with the targets. */
    isSettling(epsilon = 1e-6): boolean {
        return (
            Math.abs(this.targetPosition[0] - this.position[0]) > epsilon ||
            Math.abs(this.targetPosition[1] - this.position[1]) > epsilon ||
            Math.abs(this.targetZoom - this.zoom) > epsilon
        );
    }

    // -----------------------------------------------------------------------
    // Input
    // -----------------------------------------------------------------------

    /**
     * Pan by a screen-space delta. camera.rs:162.
     *
     * Note the pan speed divides by the *current* (smoothed) zoom, not the
     * target. Faithful to the Rust: mid-zoom-animation pans are scaled by where
     * the animation is, not where it is going.
     */
    pan(deltaX: number, deltaY: number): void {
        const panSpeed = CAMERA_DEFAULTS.panSpeed / this.zoom;

        this.targetPosition[0] += deltaX * this.sensitivity * panSpeed;
        this.targetPosition[1] += deltaY * this.sensitivity * panSpeed;

        this.applyPositionClamp();
    }

    /**
     * Zoom about the viewport centre. camera.rs:180.
     *
     * The threshold check is not decoration: without it, a stream of tiny wheel
     * deltas would ratchet `target_zoom` by amounts too small to see while still
     * costing a uniform upload every frame. It is relative so it behaves the
     * same at zoom 0.005 as at zoom 50.
     */
    zoomBy(delta: number): void {
        const zoomFactor = 1.0 + delta * this.sensitivity * CAMERA_DEFAULTS.zoomStep;
        const newZoom = this.targetZoom * zoomFactor;
        const clampedZoom = clamp(newZoom, CAMERA_DEFAULTS.minZoom, CAMERA_DEFAULTS.maxZoom);

        const threshold = Math.max(
            this.targetZoom * CAMERA_DEFAULTS.zoomRelativeThreshold,
            CAMERA_DEFAULTS.zoomAbsoluteThreshold
        );

        if (Math.abs(clampedZoom - this.targetZoom) > threshold) {
            // The Rust saved and restored target_position around this line. That
            // was a no-op — nothing between the two touches it — so it is dropped.
            this.targetZoom = clampedZoom;
        }
    }

    /**
     * Zoom keeping the world point under the cursor stationary. camera.rs:206.
     *
     * @param cursorX canvas backing-store pixels (see `screenToWorld`)
     * @param cursorY canvas backing-store pixels
     *
     * The invariant this must hold is
     *
     *     screenToWorld(cursor)  before  ==  screenToWorld(cursor)  after
     *
     * which, since `world = ndc / zoom + position`, requires
     *
     *     position' = position + ndc * (1/zoomOld - 1/zoomNew)
     *
     * **The Rust computes this with the opposite sign** (camera.rs:223: it forms
     * `(mouse_ndc - new_ndc) / target_zoom`, which expands to
     * `ndc * (1/zoomNew - 1/zoomOld)`), so the desktop app actually pushes the
     * cursor's world point *away* under the pointer, at double the error. The
     * correct sign is used here — reproducing the bug would make the stated
     * invariant untestable and the feature wrong. Everything else about the
     * function, including doing nothing when `zoomBy` was below threshold or
     * clamped, is faithful.
     */
    zoomToCursor(delta: number, cursorX: number, cursorY: number): void {
        const oldZoom = this.targetZoom;
        this.zoomBy(delta);
        const newZoom = this.targetZoom;

        // Guard the divisions: a zero-sized viewport (canvas not laid out yet)
        // would otherwise poison target_position with NaN permanently.
        if (
            !(newZoom > 0) ||
            !(oldZoom > 0) ||
            this.viewportWidth <= 0 ||
            this.viewportHeight <= 0
        ) {
            return;
        }

        const ndc = this.screenToNdc(screenCoords(cursorX, cursorY));

        const offsetX = ndc.x * (1 / oldZoom - 1 / newZoom);
        const offsetY = ndc.y * (1 / oldZoom - 1 / newZoom);

        this.targetPosition[0] += offsetX;
        this.targetPosition[1] += offsetY;

        // camera.rs:230. Note the clamp can break the stationarity invariant at
        // the edge of the pan box; that is intended — the bound wins.
        this.applyPositionClamp();
    }

    /** camera.rs:237 — snaps both current and target, so there is no glide home. */
    reset(): void {
        this.position = [0, 0];
        this.targetPosition = [0, 0];
        this.zoom = 1;
        this.targetZoom = 1;
    }

    /**
     * camera.rs:246.
     *
     * @param width  canvas **backing-store** width (`canvas.width`)
     * @param height canvas **backing-store** height (`canvas.height`)
     */
    resize(width: number, height: number): void {
        this.viewportWidth = width;
        this.viewportHeight = height;
    }

    // -----------------------------------------------------------------------
    // Coordinate conversions — all read target position/zoom, per camera.rs:278
    // -----------------------------------------------------------------------

    /**
     * Canvas backing-store pixels -> world.
     *
     * **Input is canvas backing-store pixels** — `canvas.width`/`canvas.height`
     * space, origin at the canvas's top-left corner. It is never CSS pixels,
     * never `clientX`/`clientY`, and never `window.innerWidth`-relative. Get
     * there with `clientToCanvasPx()` from `$lib/engine/gpu/pointer`; and
     * `resize()` must be fed `canvas.width`/`canvas.height` for this to line up.
     */
    screenToWorld(screen: ScreenCoords): WorldCoords {
        const ndc = this.screenToNdc(screen);
        return this.ndcToWorld(ndc);
    }

    /** World -> canvas backing-store pixels. Exact inverse of `screenToWorld`. */
    worldToScreen(world: WorldCoords): ScreenCoords {
        return this.ndcToScreen(this.worldToNdc(world));
    }

    /** camera.rs:297. Y is flipped: screen grows downward, NDC grows upward. */
    screenToNdc(screen: ScreenCoords): NdcCoords {
        const ndcX = (screen.x / this.viewportWidth) * 2.0 - 1.0;
        const ndcY = -((screen.y / this.viewportHeight) * 2.0 - 1.0);
        return ndcCoords(ndcX, ndcY);
    }

    /** camera.rs:304. */
    ndcToScreen(ndc: NdcCoords): ScreenCoords {
        return screenCoords(
            (ndc.x + 1.0) * this.viewportWidth * 0.5,
            (-ndc.y + 1.0) * this.viewportHeight * 0.5
        );
    }

    /** camera.rs:311. */
    ndcToWorld(ndc: NdcCoords): WorldCoords {
        return worldCoords(
            ndc.x / this.targetZoom + this.targetPosition[0],
            ndc.y / this.targetZoom + this.targetPosition[1]
        );
    }

    /** camera.rs:318. Matches the transform matrix the shaders use. */
    worldToNdc(world: WorldCoords): NdcCoords {
        return ndcCoords(
            (world.x - this.targetPosition[0]) * this.targetZoom,
            (world.y - this.targetPosition[1]) * this.targetZoom
        );
    }

    // -----------------------------------------------------------------------
    // State
    // -----------------------------------------------------------------------

    /** The `CameraState` contract from engine/types.ts. Reports smoothed values. */
    getState(): CameraState {
        return { position: [this.position[0], this.position[1]], zoom: this.zoom };
    }

    /**
     * The exact JSON `get_camera_state` returned from Rust (camera.rs:325),
     * three-element position included — GrayScottMode.svelte:703 destructures
     * this shape.
     */
    getStateJson(): CameraStateJson {
        return {
            position: [this.position[0], this.position[1], 0.0],
            zoom: this.zoom,
            viewport_width: this.viewportWidth,
            viewport_height: this.viewportHeight,
            aspect_ratio: this.aspectRatio(),
        };
    }

    getTargetPosition(): [number, number] {
        return [this.targetPosition[0], this.targetPosition[1]];
    }

    getTargetZoom(): number {
        return this.targetZoom;
    }

    aspectRatio(): number {
        return this.viewportWidth / this.viewportHeight;
    }

    /** camera.rs:336. */
    setSmoothingFactor(factor: number): void {
        const [lo, hi] = CAMERA_DEFAULTS.smoothingRange;
        this.smoothingFactor = clamp(factor, lo, hi);
    }

    getSmoothingFactor(): number {
        return this.smoothingFactor;
    }

    /** camera.rs:356. */
    setSensitivity(sensitivity: number): void {
        const [lo, hi] = CAMERA_DEFAULTS.sensitivityRange;
        this.sensitivity = clamp(sensitivity, lo, hi);
    }

    getSensitivity(): number {
        return this.sensitivity;
    }

    /** camera.rs:366 — pass `null` for unbounded panning. */
    setPositionClamp(range: [number, number] | null): void {
        this.positionClamp = range === null ? null : [range[0], range[1]];
        this.applyPositionClamp();
    }

    getPositionClamp(): [number, number] | null {
        return this.positionClamp === null ? null : [this.positionClamp[0], this.positionClamp[1]];
    }

    // -----------------------------------------------------------------------
    // GPU
    // -----------------------------------------------------------------------

    /**
     * Pack `CameraUniform` (camera.rs:11): mat4x4 + vec2 position + zoom +
     * aspect_ratio = 20 floats / 80 bytes, matching the WGSL struct declared in
     * e.g. slime_mold/shaders/quad.wgsl:3.
     *
     * Derived on demand rather than cached. The Rust cached it and refreshed in
     * `update`/`reset`/`resize`, but every input is a field of this object and
     * `pan`/`zoom` only touch targets, so on-demand is equivalent with one less
     * thing to keep in sync.
     */
    getUniformData(out?: Float32Array<ArrayBuffer>): Float32Array<ArrayBuffer> {
        const data = out ?? new Float32Array(CAMERA_UNIFORM_FLOATS);

        // Column-major orthographic transform, camera.rs:111. Scale about the
        // origin, then translate so the camera centre lands on NDC (0,0).
        data.fill(0);
        data[0] = this.zoom;
        data[5] = this.zoom;
        data[10] = 1.0;
        data[12] = -this.position[0] * this.zoom;
        data[13] = -this.position[1] * this.zoom;
        data[15] = 1.0;

        data[16] = this.position[0];
        data[17] = this.position[1];
        data[18] = this.zoom;
        data[19] = this.aspectRatio();

        return data;
    }

    /** Create the uniform buffer. Separate from the constructor so the camera is usable without a GPU. */
    attachToDevice(device: GPUDevice, label = 'Camera Uniform Buffer'): GPUBuffer {
        this.buffer?.destroy();
        this.buffer = device.createBuffer({
            label,
            size: CAMERA_UNIFORM_BYTES,
            usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
        });
        return this.buffer;
    }

    /** camera.rs:264. No-op if no buffer has been attached. */
    uploadToGpu(queue: GPUQueue): void {
        if (!this.buffer) return;
        queue.writeBuffer(this.buffer, 0, this.getUniformData());
    }

    /** The buffer to bind, or null before `attachToDevice`. */
    getBuffer(): GPUBuffer | null {
        return this.buffer;
    }

    /** Idempotent. */
    destroy(): void {
        this.buffer?.destroy();
        this.buffer = null;
    }

    // -----------------------------------------------------------------------

    private applyPositionClamp(): void {
        if (!this.positionClamp) return;
        const [min, max] = this.positionClamp;
        this.targetPosition[0] = clamp(this.targetPosition[0], min, max);
        this.targetPosition[1] = clamp(this.targetPosition[1], min, max);
    }
}
